drivers/at_cmd/at_cmd.c

```diff
@@ -170,77 +171,54 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 			goto next;
 		}
 
-		payload_len = get_return_code(buf, &ret);
+		payload_len = get_return_code(item->data, &ret);
 
 		if (ret.state != AT_CMD_NOTIFICATION) {
 			if ((response_buf_len > 0) &&
 			    (response_buf != NULL)) {
 				if (response_buf_len > payload_len) {
-					memcpy(response_buf,
-						buf, payload_len);
+					memcpy(response_buf, item->data,
+					       payload_len);
 				} else {
 					LOG_ERR("Response buffer not large "
 						"enough");
 
 					ret.code  = -EMSGSIZE;
 				}
 
-				callback         = false;
 				response_buf_len = 0;
 				response_buf     = NULL;
+
+				goto next;
 			}
 		}
 
 		if (payload_len == 0) {
 			goto next;
 		}
 
-		if (callback) {
-			at_cmd_handler_t tmp_callback;
-
-			if (ret.state == AT_CMD_NOTIFICATION) {
-				tmp_callback = notification_handler;
-			} else {
-				tmp_callback = current_cmd_handler;
-			}
-
-			if (!tmp_callback) {
-				goto next;
-			}
-
-			struct callback_work_item *item =
-			k_malloc(sizeof(struct callback_work_item));
-
-			if (!item) {
-				LOG_DBG("Failed to allocate work item");
-
-				ret.code  = -ENOMEM;
-
-				goto next;
-			}
-
-			item->callback = tmp_callback;
-
+		if (ret.state == AT_CMD_NOTIFICATION) {
+			item->callback = notification_handler;
+		} else {
+			item->callback = current_cmd_handler;
+		}
+next:
+		/* If no callback was set, free the item.
+		 * Otherwise, work queue callback will free it.
+		 */
+		if (item->callback == NULL) {
+			k_mem_slab_free(&rsp_work_items, (void **)&item);
+		} else {
 			k_work_init(&item->work, callback_worker);
-
-			memcpy(item->data, buf, payload_len);
-
 			k_work_submit(&item->work);
 		}
-next:
-		callback            = true;
-		current_cmd_handler = NULL;
-
-		if (ret.state != AT_CMD_NOTIFICATION) {
-			struct return_state_object ret_copy = {
-				.state = ret.state,
-				.code  = ret.code,
-			};
 
-			ret.code  = 0;
-			ret.state = AT_CMD_OK;
+		/* Notify back only if command was sent. */
+		if ((k_sem_count_get(&cmd_pending) == 0) &&
+		    (ret.state != AT_CMD_NOTIFICATION)) {
+			current_cmd_handler = NULL;
 
-			k_msgq_put(&return_code_msq, &ret_copy, K_FOREVER);
+			k_msgq_put(&return_code_msq, &ret, K_FOREVER);
```

Comment:
Do k_msgq_put create a copy internally? Else may the content of the ret object be changed before the message is consumed.

Reply:
Yes, it does.

```
The data item pointed at by the sending thread is copied to a waiting thread, if one exists; otherwise the item is copied to the message queue’s ring buffer, if space is available.
```

https://docs.zephyrproject.org/latest/reference/kernel/data_passing/message_queues.html#concepts

Reply:
ok, learned something today too :+1: